process/block/poolsCleaner/crossTxsPoolsCleaner.go

```diff
@@ -0,0 +1,285 @@
+	blockTransactionsPool    dataRetriever.ShardedDataCacherNotifier
+	rewardTransactionsPool   dataRetriever.ShardedDataCacherNotifier
+	unsignedTransactionsPool dataRetriever.ShardedDataCacherNotifier
+	rounder                  process.Rounder
+	shardCoordinator         sharding.Coordinator
+
+	mutMapCrossTxsRounds sync.RWMutex
+	mapCrossTxsRounds    map[string]*crossTxInfo
+}
+
+// NewCrossTxsPoolsCleaner will return a new cross txs pools cleaner
+func NewCrossTxsPoolsCleaner(
+	blockTracker BlockTracker,
+	dataPool dataRetriever.PoolsHolder,
+	rounder process.Rounder,
+	shardCoordinator sharding.Coordinator,
+) (*crossTxsPoolsCleaner, error) {
+
+	if check.IfNil(blockTracker) {
+		return nil, process.ErrNilBlockTracker
+	}
+	if check.IfNil(dataPool) {
+		return nil, process.ErrNilPoolsHolder
+	}
+	if check.IfNil(dataPool.Transactions()) {
+		return nil, process.ErrNilTransactionPool
+	}
+	if check.IfNil(dataPool.RewardTransactions()) {
+		return nil, process.ErrNilRewardTxDataPool
+	}
+	if check.IfNil(dataPool.UnsignedTransactions()) {
+		return nil, process.ErrNilUnsignedTxDataPool
+	}
+	if check.IfNil(rounder) {
+		return nil, process.ErrNilRounder
+	}
+	if check.IfNil(shardCoordinator) {
+		return nil, process.ErrNilShardCoordinator
+	}
+
+	ctpc := crossTxsPoolsCleaner{
+		blockTracker:             blockTracker,
+		blockTransactionsPool:    dataPool.Transactions(),
+		rewardTransactionsPool:   dataPool.RewardTransactions(),
+		unsignedTransactionsPool: dataPool.UnsignedTransactions(),
+		rounder:                  rounder,
+		shardCoordinator:         shardCoordinator,
+	}
+
+	ctpc.mapCrossTxsRounds = make(map[string]*crossTxInfo)
+
+	ctpc.blockTransactionsPool.RegisterHandler(ctpc.receivedBlockTx)
+	ctpc.rewardTransactionsPool.RegisterHandler(ctpc.receivedRewardTx)
+	ctpc.unsignedTransactionsPool.RegisterHandler(ctpc.receivedUnsignedTx)
+
+	return &ctpc, nil
+}
+
+func (ctpc *crossTxsPoolsCleaner) receivedBlockTx(key []byte) {
+	if key == nil {
+		return
+	}
+
+	log.Trace("crossTxsPoolsCleaner.receivedBlockTx", "hash", key)
+	ctpc.processReceivedTx(key, blockTx)
+}
+
+func (ctpc *crossTxsPoolsCleaner) receivedRewardTx(key []byte) {
+	if key == nil {
+		return
+	}
+
+	log.Trace("crossTxsPoolsCleaner.receivedRewardTx", "hash", key)
+	ctpc.processReceivedTx(key, rewardTx)
+}
+
+func (ctpc *crossTxsPoolsCleaner) receivedUnsignedTx(key []byte) {
+	if key == nil {
+		return
+	}
+
+	log.Trace("crossTxsPoolsCleaner.receivedUnsignedTx", "hash", key)
+	ctpc.processReceivedTx(key, unsignedTx)
+}
+
+func (ctpc *crossTxsPoolsCleaner) processReceivedTx(key []byte, txType int8) {
+	isCrossTx := false
+	var crossTxInfo *crossTxInfo
+	for shardID := uint32(0); shardID < ctpc.shardCoordinator.NumberOfShards(); shardID++ {
+		isCrossTx, crossTxInfo = ctpc.isCrossTxFromShard(key, shardID, txType)
+		if isCrossTx {
+			break
+		}
+	}
+
+	if !isCrossTx {
+		isCrossTx, crossTxInfo = ctpc.isCrossTxFromShard(key, core.MetachainShardId, txType)
```

Comment:
add a nil check for crossTxInfo as GoLand raises warnings

Reply:
Done